### ISSUE TITLE
Support for Go 1.12

### ISSUE THREAD

**Issue by mewmew** [gopherjs/gopherjs#887]
When trying to go get goherjs/gopherjs using Go 1.12 (one commit newer than go1.12beta1, golang/go@9ed9df6ca2), I get the following error:

```
$ go get -u github.com/gopherjs/gopherjs
# github.com/gopherjs/gopherjs/compiler
goget/src/github.com/gopherjs/gopherjs/compiler/compiler.go:20:9: undefined: ___GOPHERJS_REQUIRES_GO_VERSION_1_11___
```

As I did not find any open issues about this, I just wanted to write about the workaround for this error I'm currently using until gopherjs gets support for Go 1.12.

```
$ cd $GOPATH/src/github.com/gopherjs/gopherjs
$ git remote add myitcv https://github.com/myitcv/gopherjs
$ git fetch myitcv
$ git checkout go1.12
$ go get ./...
```

This successfully compiles GopherJS for Go 1.12.

```
$ gopherjs version
GopherJS 1.12.0
```

Cheers,
Robin

**Comment by myitcv** [comment]
@mewmew - just to point out the support on the go1.12 branch of my fork is still WIP (https://github.com/myitcv/gopherjs/pull/41)

**Comment by mewmew** [comment]
> @mewmew - just to point out the support on the go1.12 branch of my fork is still WIP (myitcv#41)

Hi @myitcv,

Thanks for the notice. I ran into some issue with runtime/value.go (something about not being able to convert `int8*` to `unsafe.Pointer`). So for now, I'm back at the stable Go 1.11 release.

Cheers,
Robin

**Comment by dmitshur** [comment]
> As I did not find any open issues about this

Yes, we should absolutely add support for Go 1.12. Thanks for opening an issue about it.

**Comment by hajimehoshi** [comment]
Looks like Go 1.12 was released: https://golang.org/dl/

**Comment by hajimehoshi** [comment]
@dmitshur I was wondering what is going on this...

**Comment by dmitshur** [comment]
@hajimehoshi From what I can see, no one has started work on Go 1.12 support in the main GopherJS repository yet. In https://github.com/gopherjs/gopherjs/issues/894#issuecomment-465529028, @myitcv stated his intention to start upstreaming changes from his fork into this repository, which I understand has some Go 1.12 support (e.g., PR https://github.com/myitcv/gopherjs/pull/41).

I propose we follow the same approach as for previous releases:

1. make a `go1.12` branch off of the current `master` for development
2. create first commit that targets Go 1.12 and updates GopherJS version string to "1.12-wip", analogously to commit https://github.com/gopherjs/gopherjs/pull/853/commits/061c27b405ef571c34f17b87bb0f16bb95dce7a3
3. keep adding logical commits that resolve build failures and move us closer to all tests passing, with commit messages that explain what happened
4. make a PR (analogously to PR #853), get it reviewed and merged, changing the version string to "1.12"
5. delete `go1.12` branch and create `go1.11` branch off the latest `master` before the 1.12 PR is merged

Those who don't have push rights to the main repository but wish to contribute can send PRs into the `go1.12` branch, after it is created.

Anyone who has time to work on this can create the `go1.12` branch and start pushing to it. Just try to make the commits logical and easier to review.

**Comment by hajimehoshi** [comment]
Sure, thank you for explanation! I'll try to find time to make changes for 1.12.

I think creating the branch of @myitcv on this repository should be done first. Any thoughts?

**Comment by dmitshur** [comment]
> I think creating the branch based on the branch of @myitcv on this repository should be done first.

Sorry, I don't understand what branches you're referring to. Can you clarify please?

**Comment by hajimehoshi** [comment]
I meant to adopt https://github.com/myitcv/gopherjs/pull/41 instead of creating a new go1.12 branch from scratch.

**Comment by hajimehoshi** [comment]
As we (Paul and I) discussed at Slack, as that branch comes from Paul's master branch, which is much different from GopherJS's master, I think I'll create a new branch first.

**Comment by hajimehoshi** [comment]
@myitcv noted:

>>>
To get going:

• modify `compiler/version_check.go`
• modify `compiler/compiler.go` to refer to the new version const
• Install `gopherjs`
• Start running tests via `gopherjs test --minify -v --short github.com/gopherjs/gopherjs/tests/... $(go list std | grep -v -x -f .std_test_pkg_exclusions)` (that file exists in my fork)
• Fix any issues as you come across them (edited)

**Comment by hajimehoshi** [comment]
Started working on https://github.com/gopherjs/gopherjs/tree/go1.12

**Comment by hajimehoshi** [comment]
* I needed to fix `compiler/natives/src/internal/cpu/cpu_js.go`. In this case, I'd need to `go generate ./...`. Is this correct?
* `go generate` failed like this:

```
$ go generate ./...
writing fs_vfsdata.go
writing fs_vfsdata.go
2019/03/02 17:44:14 failed to run /Users/hajimehoshi/go/src/github.com/gopherjs/gopherjs/node_modules/.bin/uglifyjs --config-file /Users/hajimehoshi/go/src/github.com/gopherjs/gopherjs/compiler/prelude/uglifyjs_options.json: fork/exec /Users/hajimehoshi/go/src/github.com/gopherjs/gopherjs/node_modules/.bin/uglifyjs: no such file or directory

exit status 1
compiler/prelude/prelude.go:3: running "go": exit status 1
```

I've already done `npm install`. Is there anything I am missing?

**Comment by dmitshur** [comment]
@hajimehoshi The first commit 852259dc2169a1ba8d093b86f1b1a8aab1be3686 looks great! Nice work!

> I needed to fix `compiler/natives/src/internal/cpu/cpu_js.go`. In this case, I'd need to `go generate ./...`. Is this correct?

The documentation for code generation is available at https://github.com/gopherjs/gopherjs/wiki/Developer-Guidelines#code-generation.

It's important to note that when developing GopherJS locally, it can be very convenient to use the `gopherjsdev` build tag. It exists specifically to avoid having to re-generate `compiler/natives` after each code change. It's documented in the 2nd last paragraph of the "Code generation" section.

> `fork/exec /Users/hajimehoshi/go/src/github.com/gopherjs/gopherjs/node_modules/.bin/uglifyjs: no such file or directory`

It sounds like `uglifyjs` did not get installed, or got installed but not in the right place. Did you run `npm install` at the root of the GopherJS directory? What was its output? /cc @myitcv

**Comment by dmitshur** [comment]
@hajimehoshi I can't reproduce the `uglifyjs` error. It works for me. Here's what I did:

```
gopherjs $ go list
github.com/gopherjs/gopherjs
gopherjs $ rm -rf ./node_modules/
gopherjs $ npm install
npm WARN gopherjs@ No repository field.
npm WARN gopherjs@ No license field.

added 3 packages from 38 contributors and audited 3 packages in 0.382s
found 0 vulnerabilities

gopherjs $ go generate ./...
writing fs_vfsdata.go
writing fs_vfsdata.go
gopherjs $ echo $?
0
```

**Comment by hajimehoshi** [comment]
Hi, I'm not sure the difference. I am of course using Go 1.12, but I don't think this is the cause:

```
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/hajimehoshi/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/hajimehoshi/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/ht/ky_bwgzs4bd5z1hh02k34x_h0000gn/T/go-build238319683=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

> Did you run npm install at the root of the GopherJS directory?

Yes I did.

```
hajimehoshi@Hajimes-MacBook-Pro ~/gopherjs
$ npm install
npm WARN gopherjs@ No repository field.
npm WARN gopherjs@ No license field.

removed 1 package and audited 3 packages in 0.355s
found 0 vulnerabilities

hajimehoshi@Hajimes-MacBook-Pro ~/gopherjs
$ npm install
npm WARN gopherjs@ No repository field.
npm WARN gopherjs@ No license field.

audited 3 packages in 0.361s
found 0 vulnerabilities
```

**Comment by hajimehoshi** [comment]
>  it can be very convenient to use the gopherjsdev build tag.

Aha ok, with the tag, it looks like I could make a progress. Thank you!

```
go run -tags=gopherjsdev . test --minify -v --short github.com/gopherjs/gopherjs/tests
```

**Comment by dmitshur** [comment]
What version of `npm`? Mine is `6.7.0`.

What's the contents of your `node_modules` after you run `npm install`? Here are mine, for comparison:

<details><br>

```
gopherjs $ rm -rf ./node_modules
gopherjs $ npm -version
6.7.0
gopherjs $ npm install
npm WARN gopherjs@ No repository field.
npm WARN gopherjs@ No license field.

added 3 packages from 38 contributors and audited 3 packages in 0.837s
found 0 vulnerabilities

gopherjs $ tree ./node_modules
./node_modules
├── commander
│   ├── CHANGELOG.md
│   ├── LICENSE
│   ├── Readme.md
│   ├── index.js
│   ├── package.json
│   └── typings
│       └── index.d.ts
├── source-map
│   ├── CHANGELOG.md
│   ├── LICENSE
│   ├── README.md
│   ├── dist
│   │   ├── source-map.debug.js
│   │   ├── source-map.js
│   │   ├── source-map.min.js
│   │   └── source-map.min.js.map
│   ├── lib
│   │   ├── array-set.js
│   │   ├── base64-vlq.js
│   │   ├── base64.js
│   │   ├── binary-search.js
│   │   ├── mapping-list.js
│   │   ├── quick-sort.js
│   │   ├── source-map-consumer.js
│   │   ├── source-map-generator.js
│   │   ├── source-node.js
│   │   └── util.js
│   ├── package.json
│   ├── source-map.d.ts
│   └── source-map.js
└── uglify-es
    ├── LICENSE
    ├── README.md
    ├── bin
    │   └── uglifyjs
    ├── lib
    │   ├── ast.js
    │   ├── compress.js
    │   ├── minify.js
    │   ├── mozilla-ast.js
    │   ├── output.js
    │   ├── parse.js
    │   ├── propmangle.js
    │   ├── scope.js
    │   ├── sourcemap.js
    │   ├── transform.js
    │   └── utils.js
    ├── package.json
    └── tools
        ├── domprops.json
        ├── exit.js
        ├── exports.js
        ├── node.js
        └── props.html

9 directories, 46 files

gopherjs $ ./node_modules/.bin/uglifyjs --version
uglify-es 3.3.9
```

</details>

**Comment by hajimehoshi** [comment]
Here you are:

```
hajimehoshi@Hajimes-MacBook-Pro ~/gopherjs                                                                                                                                                                                                 
$ rm -rf ./node_modules                                                                                                                                                                                                                    
hajimehoshi@Hajimes-MacBook-Pro ~/gopherjs                                                                                                                                                                                                 
$ npm version                                                                                                                                                                                                                              
{ npm: '6.8.0',                                                                                                                                                                                                                            
  ares: '1.14.0',                                                                                                                                                                                                                          
  cldr: '33.1',                                                                                                                                                                                                                            
  http_parser: '2.8.0',                                                                                                                                                                                                                    
  icu: '62.1',                                                                                                                                                                                                                             
  modules: '64',                                                                                                                                                                                                                           
  napi: '3',                                                                                                                                                                                                                               
  nghttp2: '1.34.0',                                                                                                                                                                                                                       
  node: '10.13.0',                                                                                                                                                                                                                         
  openssl: '1.1.0i',                                                                                                                                                                                                                       
  tz: '2018e',                                                                                                                                                                                                                             
  unicode: '11.0',                                                                                                                                                                                                                         
  uv: '1.23.2',                                                                                                                                                                                                                            
  v8: '6.8.275.32-node.36',                                                                                                                                                                                                                
  zlib: '1.2.11' }                                                                                                                                                                                                                         
hajimehoshi@Hajimes-MacBook-Pro ~/gopherjs                                                                                                                                                                                                 
$ npm  -version                                                                                                                                                                                                                            
6.8.0                                                                                                                                                                                                                                      
hajimehoshi@Hajimes-MacBook-Pro ~/gopherjs                                                                                                                                                                                                 
$ npm install                                                                                                                                                                                                                              
npm WARN gopherjs@ No repository field.                                                                                                                                                                                                    
npm WARN gopherjs@ No license field.                                                                                                                                                                                                       
                                                                                                                                                                                                                                           
added 3 packages from 38 contributors and audited 3 packages in 0.372s                                                                                                                                                                     
found 0 vulnerabilities  

hajimehoshi@Hajimes-MacBook-Pro ~/gopherjs                                                                                                                                                                                                 
$ tree ./node_modules                                                                                                                                                                                                                      
./node_modules                                                                                                                                                                                                                             
├── commander                                                                                                                                                                                                                              
│   ├── CHANGELOG.md                                                                                                                                                                                                                       
│   ├── LICENSE                                                                                                                                                                                                                            
│   ├── Readme.md                                                                                                                                                                                                                          
│   ├── index.js
│   ├── package.json
│   └── typings
│       └── index.d.ts
├── source-map
│   ├── CHANGELOG.md
│   ├── LICENSE
│   ├── README.md
│   ├── dist
│   │   ├── source-map.debug.js
│   │   ├── source-map.js
│   │   ├── source-map.min.js
│   │   └── source-map.min.js.map
│   ├── lib
│   │   ├── array-set.js
│   │   ├── base64-vlq.js
│   │   ├── base64.js
│   │   ├── binary-search.js
│   │   ├── mapping-list.js
│   │   ├── quick-sort.js
│   │   ├── source-map-consumer.js
│   │   ├── source-map-generator.js
│   │   ├── source-node.js
│   │   └── util.js
│   ├── package.json
│   ├── source-map.d.ts
│   └── source-map.js
└── uglify-es
    ├── LICENSE
    ├── README.md
    ├── bin
    │   └── uglifyjs
    ├── lib
    │   ├── ast.js
    │   ├── compress.js
    │   ├── minify.js
    │   ├── mozilla-ast.js
    │   ├── output.js
    │   ├── parse.js
    │   ├── propmangle.js
    │   ├── scope.js
    │   ├── sourcemap.js
    │   ├── transform.js
    │   └── utils.js
    ├── package.json
    └── tools
        ├── domprops.json
        ├── exit.js
        ├── exports.js
        ├── node.js
        └── props.html

9 directories, 46 files
hajimehoshi@Hajimes-MacBook-Pro ~/gopherjs
$ ./node_modules/.bin/uglifyjs --version
uglify-es 3.3.9
```

**Comment by hajimehoshi** [comment]
> /Users/hajimehoshi/go/src/github.com/gopherjs/gopherjs/node_modules/.bin/uglifyjs: no such file or directory

Ugh, oh, looks like this must be executed under `$GOPATH`?

**Comment by hajimehoshi** [comment]
Yes, `npm install` at `$GOPATH/src/github.com/gopherjs/gopherjs` succeeded. Sorry for disturbing...

**Comment by hajimehoshi** [comment]
What I am working on is syscall on Darwin introduced at https://github.com/golang/go/commit/a3b01440fef3d2833909f6651455924a1c86d192

I'd need to update compiler/natives/src/syscall for Darwin. Hmm.

**Comment by hajimehoshi** [comment]
Now all the tests except for `reflect` passed: https://circleci.com/gh/gopherjs/gopherjs/1682

Yes, `reflect` is the hardest part :-)

**Comment by hajimehoshi** [comment]
Now all the tests for the stdlib passed 😄 

I'm now trying to pass all the fixedbugs test. BTW, how can I specify the fixedbugs number when  `go test ./...` under `$GOPATH/src/github.com/gopherjs/gopherjs`?

**Comment by dmitshur** [comment]
> BTW, how can I specify the fixedbugs number

You can't specify it when doing `go test ./...`. That invokes `TestGoRepositoryCompilerTests`, which just runs `go run run.go -summary`.

To run a subset of fixedbugs tests, you need to invoke [`run.go`](https://github.com/gopherjs/gopherjs/blob/d547d1d9531ed93dbdebcbff7f83e7c876a1e0ee/tests/run.go#L205-L208) manually, and specify the tests you want to run after a `--`, like so:

```
go run run.go -- fixedbugs/issue20029.go
```

When a new version of Go comes out, there are typically new tests added that you'll need to triage and likely skip, if they're not regressions in existing behavior. They don't need to be fixed as part of the updating to new Go release process, they can be fixed anytime after (by those who are interested in doing so). For past example, see https://github.com/gopherjs/gopherjs/commit/e0db4d4b62c5ed2dbf29227237a4c3222fcc1cec.

The `run.go` file comes [from the Go project](https://github.com/golang/go/blob/fc42cf8b8ccf6753c01e063ac090b2a60e70f077/test/run.go), with GopherJS-specific modifications. We should try to keep it as close with upstream as possible.

**Comment by hajimehoshi** [comment]
Thanks, I'll try it tomorrow!

**Comment by ghasemloo** [comment]
Hi,

I ran into this issue because of another project that depends on gopherjs and I am using Go1.12.
It looks like the fix is ready to go, should I expect the issue to be fixed in a few days?

Thank you.

**Comment by dmitshur** [comment]
Yes, I plan to allocate time on reviewing the PR this weekend, and I think it’s close to being mergeable.

**Comment by dmitshur** [comment]
Now that PR #900 is merged, I've created a branch named `go1.11` at commit 847fc94819f9d5a4e7154e2203b36f6dbace6f48. It was the latest commit that supported Go 1.11.x. Anyone interested in using the latest version of GopherJS 1.11-x (for Go 1.11.x) can use that branch. This is same process as we've done all previous releases. /cc @hajimehoshi

I've also updated the playground in commit https://github.com/gopherjs/gopherjs.github.io/commit/7905a53b8aef24d05ead36751a6b1131f04f86a0, wrote an [announcement medium post](https://medium.com/gopherjs/gopherjs-1-12-1-is-released-2d2a63d748b8), and made a tweet announcing the release of GopherJS 1.12-1:

https://twitter.com/GopherJS/status/1111463453129129984

So, this is all done now! 🎉 Thank you @hajimehoshi for an excellent job on the 1.12 PR, @myitcv for the initial Go 1.12 support work on his fork, and everyone else who helped!
